src/bluesearch/entrypoint/database/download.py

```diff
@@ -163,7 +165,68 @@ def run(source: str, from_month: datetime, output_dir: Path, dry_run: bool) -> i
         logger.info(f"Start downloading {source} papers.")
         download_s3_articles(bucket, url_dict, output_dir)
         return 0
+    elif source == "arxiv":
+        logger.info("Loading libraries")
+        from google.cloud.storage import Client
+
+        from bluesearch.database.download import download_gcs_blob, get_gcs_urls
+
+        client = Client.create_anonymous_client()
+        bucket = client.bucket("arxiv-dataset")
+        if from_month < datetime(2007, 4, 1):
+            logger.error(
+                "The papers from before April 2007 follow a different format "
+                "and can't be downloaded. Please contact the developers if you "
+                "need them. To proceed please re-run the command with a "
+                "different starting month."
+            )
+            return 1
+
+        logger.info("Collecting download URLs")
+        blobs_by_month = get_gcs_urls(bucket, from_month)
 
+        if dry_run:
+            print("The following items will be downloaded:")
+            for month, month_blobs in blobs_by_month.items():
+                print(f"Month: {month}")
+                for blob in month_blobs:
+                    print(blob.name)
+            return 0
+
+        def progress_info(n_jobs, n_bytes_):
+            logger.info(f"{n_jobs} download jobs submitted ({n_bytes_:,d} bytes)")
+
+        job_names = {}
+        n_blobs = 0
+        n_bytes = 0
+        # The max_workers parameter already has a reasonable default if
+        # not specified. See python docs for ThreadPoolExecutor.
+        with ThreadPoolExecutor() as executor:
+            logger.info("Submitting download jobs to workers")
+            for blob in chain(*blobs_by_month.values()):
+                future = executor.submit(
+                    download_gcs_blob,
+                    blob,
+                    output_dir,
+                    flatten=True,
+                )
+                job_names[future] = blob.name
+                n_blobs += 1
+                n_bytes += blob.size or 0
+                if n_blobs % 100 == 0:
+                    progress_info(n_blobs, n_bytes)
```

Comment:
I guess it is nice, however, again, we don't have any multi-threading logic for the other sources and the fact that we start introducing it now is not very consistent.

I guess we leave it here and have a new PR where we generalize it to all sources (and that way one can pass `--n-jobs` via the CLI

Reply:
It's really a massive speedup to use threads, I tried without before and sequential downloads are slower by a lot. So I think without this it would be way less useable.

I might be less fussed than others with the idea that everything has to be harmonized. But I'm not against following up on this and trying to implement a similar logic for other source types.

Reply:
I created an issue here #526. I am resolving this conversation for now:)